### ISSUE TITLE
production: qps stalls on adriatic

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#31823]
I deployed a new v2.1 release candidate a couple hours ago, and adriatic has had some qps stalls since (it also had some before the push, on yesterday's release candidate, too though):

<img width="930" alt="screen shot 2018-10-24 at 2 16 39 pm" src="https://user-images.githubusercontent.com/7085343/47455481-8ff48380-d797-11e8-9d1f-fa3fbf5b1e7d.png">

I was able to grab the queries that are stuck this time around. They don't change across multiple `SHOW QUERIES` calls:

```sql
root@:26257/defaultdb> select * from [show queries] order by start limit 10;
              query_id             | node_id | user_name |              start               |                                                                                                                    query                                                                                                                     | client_address  | application_name | distributed |   phase
+----------------------------------+---------+-----------+----------------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+------------------+-------------+-----------+
  1560a13726f6cf9a0000000000000006 |       6 | root      | 2018-10-24 19:12:10.809581+00:00 | UPDATE "order" SET o_carrier_id = 6 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((2, 227105), (3, 227939), (5, 227362), (8, 227767), (9, 227020), (1, 228429), (4, 226625), (6, 227548), (7, 227412), (10, 228144))) RETURNING o_d_id, o_c_id  | 127.0.0.1:57286 |                  |    false    | executing
  1560a137446e3dbc0000000000000006 |       6 | root      | 2018-10-24 19:12:11.301989+00:00 | UPDATE "order" SET o_carrier_id = 4 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((7, 227412), (8, 227767), (10, 228144), (1, 228429), (3, 227939), (4, 226625), (6, 227548), (2, 227105), (5, 227362), (9, 227020))) RETURNING o_d_id, o_c_id  | 127.0.0.1:57268 |                  |    false    | executing
  1560a137d084fc280000000000000004 |       4 | root      | 2018-10-24 19:12:13.654228+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 7)) AND (ol_o_id BETWEEN (227432 - 20) AND (227432 - 1))) AND (s_quantity < 16))   | 127.0.0.1:56480 |                  |    true     | executing
  1560a137f99ce9840000000000000005 |       5 | root      | 2018-10-24 19:12:14.343641+00:00 | UPDATE "order" SET o_carrier_id = 7 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((8, 227768), (9, 227021), (1, 228430), (3, 227940), (5, 227363), (6, 227549), (7, 227413), (10, 228145), (2, 227106), (4, 226626))) RETURNING o_d_id, o_c_id  | 127.0.0.1:34926 |                  |    false    | executing
  1560a137feac302e0000000000000004 |       4 | root      | 2018-10-24 19:12:14.428539+00:00 | UPDATE "order" SET o_carrier_id = 3 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227940), (4, 226626), (5, 227363), (10, 228145), (1, 228430), (2, 227106), (6, 227549), (7, 227413), (8, 227768), (9, 227021))) RETURNING o_d_id, o_c_id  | 127.0.0.1:56486 |                  |    false    | executing
  1560a138c42aca6a0000000000000003 |       3 | root      | 2018-10-24 19:12:17.741935+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 5)) AND (ol_o_id BETWEEN (227380 - 20) AND (227380 - 1))) AND (s_quantity < 14))   | 127.0.0.1:35326 |                  |    true     | executing
  1560a13f74022c600000000000000006 |       6 | root      | 2018-10-24 19:12:46.461861+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((2, 227115), (3, 227949), (5, 227372), (7, 227422), (8, 227777), (9, 227030), (10, 228154), (1, 228439), (4, 226635), (6, 227558))) RETURNING o_d_id, o_c_id | 127.0.0.1:57284 |                  |    false    | executing
  1560a13fb978cf620000000000000006 |       6 | root      | 2018-10-24 19:12:47.627277+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((6, 227559), (8, 227778), (9, 227031), (10, 228155), (1, 228440), (2, 227116), (3, 227950), (4, 226636), (5, 227373), (7, 227423))) RETURNING o_d_id, o_c_id | 127.0.0.1:57288 |                  |    false    | executing
  1560a1405ca733470000000000000004 |       4 | root      | 2018-10-24 19:12:50.365056+00:00 | UPDATE "order" SET o_carrier_id = 5 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227950), (4, 226636), (5, 227373), (7, 227423), (9, 227031), (10, 228155), (1, 228440), (2, 227116), (6, 227559), (8, 227778))) RETURNING o_d_id, o_c_id  | 127.0.0.1:56482 |                  |    false    | executing
  1560a1437024a1180000000000000001 |       1 | root      | 2018-10-24 19:13:03.576917+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((1, 228443), (4, 226639), (5, 227376), (6, 227562), (9, 227034), (2, 227119), (3, 227953), (7, 227426), (8, 227781), (10, 228158))) RETURNING o_d_id, o_c_id | 127.0.0.1:33552 |                  |    false    | executing
(10 rows)

Time: 8.425971ms

root@:26257/defaultdb> select * from [show queries] order by start limit 25;
              query_id             | node_id | user_name |              start               |                                                                                                                                                                                                                                                                                                  query                                                                                                                                                                                                                                                                                                  | client_address  | application_name | distributed |   phase
+----------------------------------+---------+-----------+----------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+------------------+-------------+-----------+
  1560a13726f6cf9a0000000000000006 |       6 | root      | 2018-10-24 19:12:10.809581+00:00 | UPDATE "order" SET o_carrier_id = 6 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((2, 227105), (3, 227939), (5, 227362), (8, 227767), (9, 227020), (1, 228429), (4, 226625), (6, 227548), (7, 227412), (10, 228144))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:57286 |                  |    false    | executing
  1560a137446e3dbc0000000000000006 |       6 | root      | 2018-10-24 19:12:11.301989+00:00 | UPDATE "order" SET o_carrier_id = 4 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((7, 227412), (8, 227767), (10, 228144), (1, 228429), (3, 227939), (4, 226625), (6, 227548), (2, 227105), (5, 227362), (9, 227020))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:57268 |                  |    false    | executing
  1560a137d084fc280000000000000004 |       4 | root      | 2018-10-24 19:12:13.654228+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 7)) AND (ol_o_id BETWEEN (227432 - 20) AND (227432 - 1))) AND (s_quantity < 16))                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:56480 |                  |    true     | executing
  1560a137f99ce9840000000000000005 |       5 | root      | 2018-10-24 19:12:14.343641+00:00 | UPDATE "order" SET o_carrier_id = 7 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((8, 227768), (9, 227021), (1, 228430), (3, 227940), (5, 227363), (6, 227549), (7, 227413), (10, 228145), (2, 227106), (4, 226626))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:34926 |                  |    false    | executing
  1560a137feac302e0000000000000004 |       4 | root      | 2018-10-24 19:12:14.428539+00:00 | UPDATE "order" SET o_carrier_id = 3 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227940), (4, 226626), (5, 227363), (10, 228145), (1, 228430), (2, 227106), (6, 227549), (7, 227413), (8, 227768), (9, 227021))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:56486 |                  |    false    | executing
  1560a138c42aca6a0000000000000003 |       3 | root      | 2018-10-24 19:12:17.741935+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 5)) AND (ol_o_id BETWEEN (227380 - 20) AND (227380 - 1))) AND (s_quantity < 14))                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:35326 |                  |    true     | executing
  1560a13f74022c600000000000000006 |       6 | root      | 2018-10-24 19:12:46.461861+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((2, 227115), (3, 227949), (5, 227372), (7, 227422), (8, 227777), (9, 227030), (10, 228154), (1, 228439), (4, 226635), (6, 227558))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:57284 |                  |    false    | executing
  1560a13fb978cf620000000000000006 |       6 | root      | 2018-10-24 19:12:47.627277+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((6, 227559), (8, 227778), (9, 227031), (10, 228155), (1, 228440), (2, 227116), (3, 227950), (4, 226636), (5, 227373), (7, 227423))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:57288 |                  |    false    | executing
  1560a1405ca733470000000000000004 |       4 | root      | 2018-10-24 19:12:50.365056+00:00 | UPDATE "order" SET o_carrier_id = 5 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227950), (4, 226636), (5, 227373), (7, 227423), (9, 227031), (10, 228155), (1, 228440), (2, 227116), (6, 227559), (8, 227778))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:56482 |                  |    false    | executing
  1560a1437024a1180000000000000001 |       1 | root      | 2018-10-24 19:13:03.576917+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((1, 228443), (4, 226639), (5, 227376), (6, 227562), (9, 227034), (2, 227119), (3, 227953), (7, 227426), (8, 227781), (10, 228158))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:33552 |                  |    false    | executing
  1560a143d9da68480000000000000003 |       3 | root      | 2018-10-24 19:13:05.350421+00:00 | UPDATE "order" SET o_carrier_id = 8 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227954), (4, 226640), (5, 227377), (10, 228159), (1, 228444), (2, 227120), (6, 227563), (7, 227427), (8, 227782), (9, 227035))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:35344 |                  |    false    | executing
  1560a144845952500000000000000004 |       4 | root      | 2018-10-24 19:13:08.210875+00:00 | UPDATE "order" SET o_carrier_id = 7 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((1, 228444), (3, 227954), (5, 227377), (6, 227563), (8, 227782), (9, 227035), (2, 227120), (4, 226640), (7, 227427), (10, 228159))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:56474 |                  |    false    | executing
  1560a1468ddd2cf00000000000000006 |       6 | root      | 2018-10-24 19:13:16.960368+00:00 | UPDATE "order" SET o_carrier_id = 7 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((6, 227565), (8, 227784), (9, 227037), (1, 228446), (5, 227379), (4, 226642), (7, 227429), (10, 228161), (2, 227122), (3, 227956))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:57294 |                  |    false    | executing
  1560a146958370b50000000000000006 |       6 | root      | 2018-10-24 19:13:17.088795+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227956), (4, 226642), (6, 227565), (9, 227037), (10, 228161), (1, 228446), (2, 227122), (5, 227379), (7, 227429), (8, 227784))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:57280 |                  |    false    | executing
  1560a146c6193c7e0000000000000002 |       2 | root      | 2018-10-24 19:13:17.90368+00:00  | UPDATE customer SET (c_balance, c_ytd_payment, c_payment_cnt, c_data) = (c_balance - 370.680000, c_ytd_payment + 370.680000, c_payment_cnt + 1, CASE c_credit WHEN 'BC' THEN left((((((c_id::STRING || c_d_id::STRING) || c_w_id::STRING) || 10::STRING) || 2::STRING) || 370.680000::STRING) || c_data, 500) ELSE c_data END) WHERE ((c_w_id = 9) AND (c_d_id = 1)) AND (c_id = 2400) RETURNING c_first, c_middle, c_last, c_street_1, c_street_2, c_city, c_state, c_zip, c_phone, c_since, c_credit, c_credit_lim, c_discount, c_balance, CASE c_credit WHEN 'BC' THEN left(c_data, 200) ELSE '' END | 127.0.0.1:56588 |                  |    false    | executing
  1560a146cd5774f00000000000000003 |       3 | root      | 2018-10-24 19:13:18.02542+00:00  | UPDATE "order" SET o_carrier_id = 7 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((6, 227565), (7, 227429), (8, 227784), (9, 227037), (10, 228161), (2, 227122), (3, 227956), (5, 227379), (1, 228446), (4, 226642))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:35352 |                  |    false    | executing
  1560a146d84fa2de0000000000000001 |       1 | root      | 2018-10-24 19:13:18.209458+00:00 | SELECT w_tax FROM warehouse WHERE w_id = 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:33550 |                  |    false    | executing
  1560a146db875da70000000000000002 |       2 | root      | 2018-10-24 19:13:18.263428+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 1308.050000 WHERE w_id = 2 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | 127.0.0.1:56596 |                  |    false    | executing
  1560a146de14faff0000000000000003 |       3 | root      | 2018-10-24 19:13:18.306123+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 3573.310000 WHERE w_id = 2 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | 127.0.0.1:35342 |                  |    false    | executing
  1560a146e4192aec0000000000000006 |       6 | root      | 2018-10-24 19:13:18.407267+00:00 | SELECT w_tax FROM warehouse WHERE w_id = 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:57290 |                  |    false    | executing
  1560a146fc7d137a0000000000000006 |       6 | root      | 2018-10-24 19:13:18.79744+00:00  | SELECT w_tax FROM warehouse WHERE w_id = 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:57298 |                  |    false    | executing
  1560a147021c0d8a0000000000000004 |       4 | root      | 2018-10-24 19:13:18.910703+00:00 | UPDATE "order" SET o_carrier_id = 4 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((5, 227380), (7, 227430), (2, 227123), (3, 227957), (4, 226643), (9, 227038), (10, 228162), (1, 228447), (6, 227566), (8, 227785))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                             | 127.0.0.1:56456 |                  |    false    | executing
  1560a1470a249c140000000000000005 |       5 | root      | 2018-10-24 19:13:19.04557+00:00  | SELECT w_tax FROM warehouse WHERE w_id = 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:34916 |                  |    false    | executing
  1560a1470cae87240000000000000001 |       1 | root      | 2018-10-24 19:13:19.08817+00:00  | SELECT w_tax FROM warehouse WHERE w_id = 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | 127.0.0.1:33526 |                  |    false    | executing
  1560a1470fcaaa8a0000000000000005 |       5 | root      | 2018-10-24 19:13:19.140293+00:00 | SELECT d_next_o_id FROM district WHERE (d_w_id = 2) AND (d_id = 6)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:34904 |                  |    false    | executing
(25 rows)

Time: 21.42194ms
```

There are only two "problem ranges", and they're both just non-active leases from a node's previous liveness epoch; nothing too suspicious.

I'm continuing to investigate.

**Comment by a-robinson** [comment]
It just unstuck for a little while and then got stuck again, and once again `UPDATE orders ...` queries are the oldest running queries, but once again also with some slightly less old UPDATEs on other tables stuck as well:

```sql
root@:26257/defaultdb> select * from [show queries] order by start limit 25;
              query_id             | node_id | user_name |              start               |                                                                                                                                                                                                                                                                                                 query                                                                                                                                                                                                                                                                                                 | client_address  | application_name | distributed |   phase
+----------------------------------+---------+-----------+----------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+------------------+-------------+-----------+
  1560a205ee9cfa8c0000000000000006 |       6 | root      | 2018-10-24 19:26:58.922375+00:00 | UPDATE "order" SET o_carrier_id = 10 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((2, 227126), (5, 227383), (6, 227569), (8, 227788), (1, 228450), (4, 226646), (7, 227433), (9, 227041), (10, 228165), (3, 227960))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                          | 127.0.0.1:57280 |                  |    false    | executing
  1560a207a958cbb40000000000000003 |       3 | root      | 2018-10-24 19:27:06.350215+00:00 | UPDATE "order" SET o_carrier_id = 8 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((10, 228169), (2, 227130), (3, 227964), (4, 226650), (5, 227387), (6, 227573), (8, 227792), (9, 227045), (1, 228454), (7, 227437))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                           | 127.0.0.1:35344 |                  |    false    | executing
  1560a20c036a5b7d0000000000000006 |       6 | root      | 2018-10-24 19:27:25.041138+00:00 | UPDATE "order" SET o_carrier_id = 4 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((5, 227393), (6, 227579), (8, 227798), (10, 228175), (7, 227443), (9, 227051), (1, 228460), (2, 227136), (3, 227970), (4, 226656))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                           | 127.0.0.1:57268 |                  |    false    | executing
  1560a20c040419dd0000000000000006 |       6 | root      | 2018-10-24 19:27:25.051291+00:00 | UPDATE "order" SET o_carrier_id = 6 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((1, 228460), (6, 227579), (7, 227443), (9, 227051), (10, 228175), (2, 227136), (3, 227970), (4, 226656), (5, 227393), (8, 227798))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                           | 127.0.0.1:57286 |                  |    false    | executing
  1560a20c15cef4cf0000000000000005 |       5 | root      | 2018-10-24 19:27:25.349681+00:00 | UPDATE customer SET (c_balance, c_ytd_payment, c_payment_cnt, c_data) = (c_balance - 237.300000, c_ytd_payment + 237.300000, c_payment_cnt + 1, CASE c_credit WHEN 'BC' THEN left((((((c_id::STRING || c_d_id::STRING) || c_w_id::STRING) || 7::STRING) || 9::STRING) || 237.300000::STRING) || c_data, 500) ELSE c_data END) WHERE ((c_w_id = 9) AND (c_d_id = 7)) AND (c_id = 551) RETURNING c_first, c_middle, c_last, c_street_1, c_street_2, c_city, c_state, c_zip, c_phone, c_since, c_credit, c_credit_lim, c_discount, c_balance, CASE c_credit WHEN 'BC' THEN left(c_data, 200) ELSE '' END | 127.0.0.1:34928 |                  |    false    | executing
  1560a20c18106dda0000000000000004 |       4 | root      | 2018-10-24 19:27:25.387667+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 7) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:56470 |                  |    false    | executing
  1560a20c2986f5ec0000000000000005 |       5 | root      | 2018-10-24 19:27:25.680628+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 1192.810000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:34930 |                  |    false    | executing
  1560a20c31cc46f50000000000000004 |       4 | root      | 2018-10-24 19:27:25.819435+00:00 | SELECT w_tax FROM warehouse WHERE w_id = 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:56476 |                  |    false    | executing
  1560a20c35f9bcd30000000000000003 |       3 | root      | 2018-10-24 19:27:25.889473+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 1039.310000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:35356 |                  |    false    | executing
  1560a20c3f09522e0000000000000002 |       2 | root      | 2018-10-24 19:27:26.041519+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 4152.950000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:56572 |                  |    false    | executing
  1560a20c45296ffb0000000000000001 |       1 | root      | 2018-10-24 19:27:26.144263+00:00 | SELECT w_tax FROM warehouse WHERE w_id = 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:33544 |                  |    false    | executing
  1560a20c45306e590000000000000002 |       2 | root      | 2018-10-24 19:27:26.144761+00:00 | SELECT w_tax FROM warehouse WHERE w_id = 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:56590 |                  |    false    | executing
  1560a20c4619f7300000000000000003 |       3 | root      | 2018-10-24 19:27:26.160065+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 7) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:35352 |                  |    false    | executing
  1560a20c5afaa41c0000000000000001 |       1 | root      | 2018-10-24 19:27:26.510308+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 5) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:33526 |                  |    false    | executing
  1560a20c5bb936580000000000000003 |       3 | root      | 2018-10-24 19:27:26.522789+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 1602.870000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:35348 |                  |    false    | executing
  1560a20c6fd85a630000000000000006 |       6 | root      | 2018-10-24 19:27:26.860265+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 1465.540000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:57288 |                  |    false    | executing
  1560a20c7ebf719f0000000000000003 |       3 | root      | 2018-10-24 19:27:27.109647+00:00 | UPDATE "order" SET o_carrier_id = 6 WHERE (o_w_id = 9) AND ((o_d_id, o_id) IN ((3, 227970), (4, 226656), (6, 227579), (10, 228175), (9, 227051), (1, 228460), (2, 227136), (5, 227393), (7, 227443), (8, 227798))) RETURNING o_d_id, o_c_id                                                                                                                                                                                                                                                                                                                                                           | 127.0.0.1:35338 |                  |    false    | executing
  1560a20c8e81a8280000000000000005 |       5 | root      | 2018-10-24 19:27:27.374803+00:00 | SELECT w_tax FROM warehouse WHERE w_id = 9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | 127.0.0.1:34904 |                  |    false    | executing
  1560a20c9558d8440000000000000003 |       3 | root      | 2018-10-24 19:27:27.489499+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 7) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:35350 |                  |    false    | executing
  1560a20c9685962a0000000000000002 |       2 | root      | 2018-10-24 19:27:27.509203+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 255.020000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                       | 127.0.0.1:56598 |                  |    false    | executing
  1560a20cb92ba9fd0000000000000004 |       4 | root      | 2018-10-24 19:27:28.090544+00:00 | UPDATE warehouse SET w_ytd = w_ytd + 3268.030000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:56480 |                  |    false    | executing
  1560a20cd3132e4d0000000000000004 |       4 | root      | 2018-10-24 19:27:28.52512+00:00  | UPDATE warehouse SET w_ytd = w_ytd + 2255.560000 WHERE w_id = 9 RETURNING w_name, w_street_1, w_street_2, w_city, w_state, w_zip                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:56486 |                  |    false    | executing
  1560a20cdb44e3220000000000000001 |       1 | root      | 2018-10-24 19:27:28.662659+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 9) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:33540 |                  |    false    | executing
  1560a20cedfd99ab0000000000000003 |       3 | root      | 2018-10-24 19:27:28.976725+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 2) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:35326 |                  |    false    | executing
  1560a20ceea2a7be0000000000000005 |       5 | root      | 2018-10-24 19:27:28.987564+00:00 | UPDATE district SET d_next_o_id = d_next_o_id + 1 WHERE (d_w_id = 9) AND (d_id = 7) RETURNING d_tax, d_next_o_id                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 127.0.0.1:34924 |                  |    false    | executing
(25 rows)

Time: 14.721635ms
```

**Comment by a-robinson** [comment]
Either our "Slow Requests" dashboard is broken, or this is a failure mode I've never seen before. How were there no slow requests in the system through any of this?

<img width="977" alt="screen shot 2018-10-24 at 2 34 28 pm" src="https://user-images.githubusercontent.com/7085343/47456371-efec2980-d799-11e8-93ae-117ebe0d3d22.png">

**Comment by bdarnell** [comment]
That dashboard is only tracking slowness at the KV layer (and only particular kinds of slowness). It excludes, for example, transactions that are blocking in the txnWaitQueue.

**Comment by a-robinson** [comment]
But shouldn't txns in the txnWaitQueue be blocked on a transaction that's slow at a lower level?

**Comment by bdarnell** [comment]
Not necessarily. Higher layers can contribute to poor performance too :)

**Comment by tschottdorf** [comment]
For example, a txn that's stuck restarting endlessly could also hold up others without ever spending significant time in one place.

**Comment by nvanbenschoten** [comment]
We have 4 ranges with invalid leases: `2` `606`, `661`, and `1469`.

**Comment by a-robinson** [comment]
One is just due to lack of use of an expiration-based lease (although I'm not sure why auto-renewing it didn't work), and the other three due to being at an old liveness epoch and presumably not having been needed since the epoch increment. Nothing too suspicious there.

What is odd are these distsql graphs:

<img width="932" alt="screen shot 2018-10-24 at 2 47 22 pm" src="https://user-images.githubusercontent.com/7085343/47457305-5a05ce00-d79c-11e8-8767-67c8e2bffa12.png">

What's that one stuck flow during the outage? And what does the "active queries" being equal to the number of sql connections mean? It seems as though distsql is stuck, somehow.

**Comment by tschottdorf** [comment]
Do you get any queries that actually run in `distributed` mode?

**Comment by a-robinson** [comment]
Good question, and yes, there are old stuck queries in `distributed` mode during the current stall:

```sql
root@:26257/defaultdb> select * from [show queries] where distributed = true order by start limit 25;
              query_id             | node_id | user_name |              start               |                                                                                                                    query                                                                                                                    | client_address  | application_name | distributed |   phase
+----------------------------------+---------+-----------+----------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------+------------------+-------------+-----------+
  1560a35060ac5d720000000000000003 |       3 | root      | 2018-10-24 19:50:38.175039+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 3)) AND (ol_o_id BETWEEN (228339 - 20) AND (228339 - 1))) AND (s_quantity < 12))  | 127.0.0.1:35346 |                  |    true     | executing
  1560a35094b02a470000000000000005 |       5 | root      | 2018-10-24 19:50:39.047739+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 7)) AND (ol_o_id BETWEEN (227826 - 20) AND (227826 - 1))) AND (s_quantity < 18))  | 127.0.0.1:34932 |                  |    true     | executing
  1560a3519e2257c00000000000000003 |       3 | root      | 2018-10-24 19:50:43.501246+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 10)) AND (ol_o_id BETWEEN (228571 - 20) AND (228571 - 1))) AND (s_quantity < 16)) | 127.0.0.1:35352 |                  |    true     | executing
  1560a353db24803b0000000000000004 |       4 | root      | 2018-10-24 19:50:53.114671+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 7)) AND (ol_o_id BETWEEN (227839 - 20) AND (227839 - 1))) AND (s_quantity < 11))  | 127.0.0.1:56476 |                  |    true     | executing
  1560a35a0e3935900000000000000002 |       2 | root      | 2018-10-24 19:51:19.741523+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 3)) AND (ol_o_id BETWEEN (228348 - 20) AND (228348 - 1))) AND (s_quantity < 13))  | 127.0.0.1:56594 |                  |    true     | executing
  1560a35b894cef040000000000000002 |       2 | root      | 2018-10-24 19:51:26.101392+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 5)) AND (ol_o_id BETWEEN (227786 - 20) AND (227786 - 1))) AND (s_quantity < 19))  | 127.0.0.1:56596 |                  |    true     | executing
  1560a35c859387340000000000000004 |       4 | root      | 2018-10-24 19:51:30.333888+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 3)) AND (ol_o_id BETWEEN (228350 - 20) AND (228350 - 1))) AND (s_quantity < 15))  | 127.0.0.1:56472 |                  |    true     | executing
  1560a35d23d0e6a10000000000000005 |       5 | root      | 2018-10-24 19:51:32.988708+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 8) AND (ol_d_id = 8)) AND (ol_o_id BETWEEN (227632 - 20) AND (227632 - 1))) AND (s_quantity < 20))  | 127.0.0.1:34918 |                  |    true     | executing
  1560a35d336822930000000000000004 |       4 | root      | 2018-10-24 19:51:33.250324+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 10)) AND (ol_o_id BETWEEN (228580 - 20) AND (228580 - 1))) AND (s_quantity < 20)) | 127.0.0.1:56482 |                  |    true     | executing
  1560a35df3ba26040000000000000004 |       4 | root      | 2018-10-24 19:51:36.476876+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 7)) AND (ol_o_id BETWEEN (227849 - 20) AND (227849 - 1))) AND (s_quantity < 16))  | 127.0.0.1:56474 |                  |    true     | executing
  1560a3601de552b50000000000000002 |       2 | root      | 2018-10-24 19:51:45.774283+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 8) AND (ol_d_id = 9)) AND (ol_o_id BETWEEN (228000 - 20) AND (228000 - 1))) AND (s_quantity < 14))  | 127.0.0.1:56572 |                  |    true     | executing
  1560a3615767db2b0000000000000001 |       1 | root      | 2018-10-24 19:51:51.034143+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 9) AND (ol_d_id = 9)) AND (ol_o_id BETWEEN (227455 - 20) AND (227455 - 1))) AND (s_quantity < 17))  | 127.0.0.1:33552 |                  |    true     | executing
  1560a3616fd925080000000000000006 |       6 | root      | 2018-10-24 19:51:51.444192+00:00 | SELECT count(*) FROM (SELECT DISTINCT s_i_id FROM order_line JOIN stock ON (s_i_id = ol_i_id) AND (s_w_id = ol_w_id) WHERE (((ol_w_id = 8) AND (ol_d_id = 9)) AND (ol_o_id BETWEEN (228000 - 20) AND (228000 - 1))) AND (s_quantity < 13))  | 127.0.0.1:57294 |                  |    true     | executing
(13 rows)

Time: 9.713858ms
```

**Comment by a-robinson** [comment]
None of those queries are budging.

**Comment by tschottdorf** [comment]
Not sure if relevant, but note the slight uptick in txn restarts here:

![image](https://user-images.githubusercontent.com/5076964/47457618-de297700-d7d7-11e8-82ed-2a3156006ffc.png)

**Comment by bdarnell** [comment]
If you look back at past stalls, the "active flows" graph is stable in all of them, but it's not always showing a single flow. There was a stall with no flows, and one with three. I think these flows are a symptom of the issue rather than the cause.

**Comment by nvanbenschoten** [comment]
Goroutine dumps were very helpful in debugging stalls with roachtests. Let's take a look at those.

**Comment by a-robinson** [comment]
Perhaps, yeah. This time there are 13 stuck distsql queries and 13 stuck flows.

**Comment by tschottdorf** [comment]
https://cockroach-adriatic-0006.crdb.io:8080/debug/logspy?count=100&duration=10s&grep=/Table/57/1/9/8/-228187/0

```
I181024 19:59:05.522947 7482778 server/debug/logspy.go:182  intercepting logs with options {Count:100 Duration:10s Grep:/Table/57/1/9/8/-228187/0}
I181024 19:59:05.584473 7375389 storage/replica_command.go:112  [n6,s6,r1120/1:/Table/57/1/9/{7/-60…-10/-4…}] evaluated Put command header:<key:"\301\211\221\220\205\374\204\245\210" sequence:43 > value:<raw_bytes:"z\356\013\010\nC\266'\030\376\257\206\275\013\000\023\n\023\022\023\002" timestamp:<> > : header:<> , err=conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.584586 7375389 storage/intent_resolver.go:319  [n6,s6] 862a79b4 encountered another write intent error conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.584602 7375389 storage/intent_resolver.go:416  [n6,s6] resolving write intent conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.584617 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 19:59:05.738382 7375389 storage/replica_command.go:112  [n6,s6,r1120/1:/Table/57/1/9/{7/-60…-10/-4…}] evaluated Put command header:<key:"\301\211\221\220\205\374\204\245\210" sequence:43 > value:<raw_bytes:"z\356\013\010\nC\266'\030\376\257\206\275\013\000\023\n\023\022\023\002" timestamp:<> > : header:<> , err=conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.738478 7375389 storage/intent_resolver.go:319  [n6,s6] 862a79b4 encountered another write intent error conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.738493 7375389 storage/intent_resolver.go:416  [n6,s6] resolving write intent conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.738508 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 19:59:05.891025 7375389 storage/replica_command.go:112  [n6,s6,r1120/1:/Table/57/1/9/{7/-60…-10/-4…}] evaluated Put command header:<key:"\301\211\221\220\205\374\204\245\210" sequence:43 > value:<raw_bytes:"z\356\013\010\nC\266'\030\376\257\206\275\013\000\023\n\023\022\023\002" timestamp:<> > : header:<> , err=conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.891126 7375389 storage/intent_resolver.go:319  [n6,s6] 862a79b4 encountered another write intent error conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.891142 7375389 storage/intent_resolver.go:416  [n6,s6] resolving write intent conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:05.891158 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 19:59:06.044938 7375389 storage/replica_command.go:112  [n6,s6,r1120/1:/Table/57/1/9/{7/-60…-10/-4…}] evaluated Put command header:<key:"\301\211\221\220\205\374\204\245\210" sequence:43 > value:<raw_bytes:"z\356\013\010\nC\266'\030\376\257\206\275\013\000\023\n\023\022\023\002" timestamp:<> > : header:<> , err=conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:06.045035 7375389 storage/intent_resolver.go:319  [n6,s6] 862a79b4 encountered another write intent error conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:06.045063 7375389 storage/intent_resolver.go:416  [n6,s6] resolving write intent conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:06.045088 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 19:59:06.198468 7375389 storage/replica_command.go:112  [n6,s6,r1120/1:/Table/57/1/9/{7/-60…-10/-4…}] evaluated Put command header:<key:"\301\211\221\220\205\374\204\245\210" sequence:43 > value:<raw_bytes:"z\356\013\010\nC\266'\030\376\257\206\275\013\000\023\n\023\022\023\002" timestamp:<> > : header:<> , err=conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:06.198549 7375389 storage/intent_resolver.go:319  [n6,s6] 862a79b4 encountered another write intent error conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:06.198565 7375389 storage/intent_resolver.go:416  [n6,s6] resolving write intent conflicting intents on /Table/57/1/9/8/-228187/0
I181024 19:59:06.198580 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
...
```

**Comment by tschottdorf** [comment]
https://cockroach-adriatic-0006.crdb.io:8080/debug/logspy?count=100&duration=10s&grep=1c11ebbf

```
I181024 20:00:24.601394 7482778 server/debug/logspy.go:182  intercepting logs with options {Count:100 Duration:10s Grep:1c11ebbf}
E181024 20:00:24.632503 7375389 storage/intent_resolver.go:303  [n6,s6] while waiting in push contention queue to push 1c11ebbf: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" id=862a79b4 key=/Table/57/1/9/1/-228849/0 rw=true pri=0.04666341 iso=SERIALIZABLE stat=ABORTED epo=1 ts=1540410639.803767105,0 orig=1540410639.803767105,0 max=1540410633.292410802,0 wto=false rop=false seq=51
I181024 20:00:24.633222 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 20:00:24.733521 7375389 storage/intent_resolver.go:301  [n6,s6] 862a79b4 pushing 1c11ebbf to detect dependency cycles
E181024 20:00:24.785548 7375389 storage/intent_resolver.go:303  [n6,s6] while waiting in push contention queue to push 1c11ebbf: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" id=862a79b4 key=/Table/57/1/9/1/-228849/0 rw=true pri=0.04666341 iso=SERIALIZABLE stat=ABORTED epo=1 ts=1540410639.803767105,0 orig=1540410639.803767105,0 max=1540410633.292410802,0 wto=false rop=false seq=51
I181024 20:00:24.785989 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 20:00:24.886213 7375389 storage/intent_resolver.go:301  [n6,s6] 862a79b4 pushing 1c11ebbf to detect dependency cycles
E181024 20:00:24.938779 7375389 storage/intent_resolver.go:303  [n6,s6] while waiting in push contention queue to push 1c11ebbf: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" id=862a79b4 key=/Table/57/1/9/1/-228849/0 rw=true pri=0.04666341 iso=SERIALIZABLE stat=ABORTED epo=1 ts=1540410639.803767105,0 orig=1540410639.803767105,0 max=1540410633.292410802,0 wto=false rop=false seq=51
I181024 20:00:24.939551 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 20:00:25.039835 7375389 storage/intent_resolver.go:301  [n6,s6] 862a79b4 pushing 1c11ebbf to detect dependency cycles
E181024 20:00:25.092230 7375389 storage/intent_resolver.go:303  [n6,s6] while waiting in push contention queue to push 1c11ebbf: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" id=862a79b4 key=/Table/57/1/9/1/-228849/0 rw=true pri=0.04666341 iso=SERIALIZABLE stat=ABORTED epo=1 ts=1540410639.803767105,0 orig=1540410639.803767105,0 max=1540410633.292410802,0 wto=false rop=false seq=51
I181024 20:00:25.092826 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 20:00:25.193111 7375389 storage/intent_resolver.go:301  [n6,s6] 862a79b4 pushing 1c11ebbf to detect dependency cycles
E181024 20:00:25.244917 7375389 storage/intent_resolver.go:303  [n6,s6] while waiting in push contention queue to push 1c11ebbf: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" id=862a79b4 key=/Table/57/1/9/1/-228849/0 rw=true pri=0.04666341 iso=SERIALIZABLE stat=ABORTED epo=1 ts=1540410639.803767105,0 orig=1540410639.803767105,0 max=1540410633.292410802,0 wto=false rop=false seq=51
I181024 20:00:25.245388 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 20:00:25.345646 7375389 storage/intent_resolver.go:301  [n6,s6] 862a79b4 pushing 1c11ebbf to detect dependency cycles
E181024 20:00:25.400505 7375389 storage/intent_resolver.go:303  [n6,s6] while waiting in push contention queue to push 1c11ebbf: TransactionAbortedError(ABORT_REASON_PUSHER_ABORTED): "sql txn" id=862a79b4 key=/Table/57/1/9/1/-228849/0 rw=true pri=0.04666341 iso=SERIALIZABLE stat=ABORTED epo=1 ts=1540410639.803767105,0 orig=1540410639.803767105,0 max=1540410633.292410802,0 wto=false rop=false seq=51
I181024 20:00:25.400873 7375389 storage/intent_resolver.go:188  [n6,s6] adding 862a79b4 to contention queue on intent /Table/57/1/9/8/-228187/0 @1c11ebbf
I181024 20:00:25.501131 7375389 storage/intent_resolver.go:301  [n6,s6] 862a79b4 pushing 1c11ebbf to detect dependency cycles
...
```

Maybe this is all in good working order, but it doesn't seem good.

**Comment by tschottdorf** [comment]
(We really need better insight into the contention queue, at the very least a few metrics that likely would've spiked during this event)

**Comment by a-robinson** [comment]
> (We really need better insight into the contention queue, at the very least a few metrics that likely would've spiked during this event)

+1

**Comment by bdarnell** [comment]
#31106. Debuggability needs to be a prerequisite for merging any new component like this.

**Comment by bdarnell** [comment]
Repeated `ABORT_REASON_PUSHER_ABORTED` for the same transaction is not normal. This means that there's apparently been a deadlock, and txn 862a79b4 has been aborted but doesn't know it yet. The txnwait queue is discovering this fact, but it's not filtering back up to the TxnCoordSender (and the TxnCoordSender isn't failing its heartbeats?). This transaction is therefore busy-looping, and probably blocking other transactions that would be able to resolve the intent here.

**Comment by bdarnell** [comment]
Not as pathological as the repeated ABORT_REASON_PUSHER_ABORTED, but I see a few things that look suboptimal at least.

```
I181024 21:09:46.020024 12675247 storage/intent_resolver.go:257  [n1,s1] ddea36a9 exiting contention queue to proceed
I181024 21:09:46.020633 12675246 storage/intent_resolver.go:244  [n1,s1] 394dca99 exiting contention queue to push ddea36a9
```

This pattern (one txn "exiting to proceed", then a second txn that exits to push the one that's "proceeding") recurs a few times and seems suboptimal. Are txns getting knocked out as soon as they exit the queue and are unable to proceed? 

We still see some small thundering herds. Here 6 txns "exit to proceed" and 3 of them go right back in the queue, on the same intent.

```
I181024 21:09:46.304272 14912525 storage/intent_resolver.go:257  [n3,s3] 241050c9 exiting contention queue to proceed
I181024 21:09:46.310395 14912778 storage/intent_resolver.go:257  [n3,s3] 9cb1ffa0 exiting contention queue to proceed
I181024 21:09:46.312102 14913156 storage/intent_resolver.go:257  [n3,s3] ce0b9dea exiting contention queue to proceed
I181024 21:09:46.312601 14913186 storage/intent_resolver.go:257  [n3,s3] 609482f3 exiting contention queue to proceed
I181024 21:09:46.313480 14913380 storage/intent_resolver.go:257  [n3,s3] ddea36a9 exiting contention queue to proceed
I181024 21:09:46.314830 14913432 storage/intent_resolver.go:257  [n3,s3] ae94513a exiting contention queue to proceed
I181024 21:09:46.323240 14913432 storage/intent_resolver.go:188  [n3,s3] adding ae94513a to contention queue on intent /Table/53/1/7/0 @241050c9
I181024 21:09:46.324786 14913536 storage/intent_resolver.go:188  [n3,s3] adding 9cb1ffa0 to contention queue on intent /Table/53/1/7/0 @241050c9
I181024 21:09:46.324823 14912274 storage/intent_resolver.go:244  [n3,s3] 4b61822a exiting contention queue to push 241050c9
I181024 21:09:46.325955 14913537 storage/intent_resolver.go:188  [n3,s3] adding ce0b9dea to contention queue on intent /Table/53/1/7/0 @241050c9
I181024 21:09:46.336760 14913463 storage/intent_resolver.go:257  [n3,s3] 465949c4 exiting contention queue to proceed
I181024 21:09:46.337086 14913463 storage/intent_resolver.go:188  [n3,s3] adding 465949c4 to contention queue on intent /Table/53/1/7/0 @241050c9
I181024 21:09:46.337117 14913432 storage/intent_resolver.go:244  [n3,s3] ae94513a exiting contention queue to push 241050c9
```

**Comment by bdarnell** [comment]
> I deployed a new v2.1 release candidate a couple hours ago, and adriatic has had some qps stalls since (it also had some before the push, on yesterday's release candidate, too though):

Because of the late change to disable the merge queue, rc1 and rc2 were pretty close together. Deployment logs show that we pushed 670c7d3df7 on Oct 16 23:00, de005ed7e on Oct 23 23:03, and 0dba537ae8 on Oct 24 17:27. That Oct 23 push lines up with the first observation of this bug from before the rc2 push. That means that instead of lying dormant for an unknown number of weeks before being observed, the problem here was very likely created by some change [between these two commits](https://github.com/cockroachdb/cockroach/compare/670c7d3df74e0efbc633ceac8e97fada3b1063b2...de005ed7e08f4edd3b95b9388596e149058d13b8).

**Comment by bdarnell** [comment]
> I don't know what to make of this. Scans are repeatedly missing intents that RefreshRanges over the same spans are seeing - that's very strange.

My guess is that these are the same intents, but the timestamps are being reported differently. The timestamp is stored twice: once in the intent's MVCCKey, and once in the TxnMeta stored in the MVCCMetadata. If those somehow differed, I think we could see oddities like this.

**Comment by bdarnell** [comment]
> If those somehow differed, I think we could see oddities like this.

This "somehow" is not actually unusual. A transaction that has had its timestamp pushed and successfully refreshed will write its intents using MVCC keys at the RefreshedTimestamp, but it does not appear to update its TxnMeta.Timestamp until commit time (or maybe it does and I haven't found it yet). The relationship between these timestamps is confusing. I think in the debugging snippets above it's just a case of different timestamps being logged by different pieces of code, but I wouldn't be at all surprised if something was looking at the wrong one and leading to a near-deadlock here.

**Comment by nvanbenschoten** [comment]
I think I confused you. I wasn't saying that I saw an intent at the wrong timestamp, I just didn't see the intent at all when running `cockroach debug keys`.

**Comment by bdarnell** [comment]
OK. RefreshRange uses a timebound iterator, which I could see potentially causing it to miss things, but I don't see how that could make it see something that the simpler `debug keys` doesn't (unless there's bad iterator/snapshot caching?)

**Comment by nvanbenschoten** [comment]
A bad iterator/snapshot caching issue would explain what I'm seeing.

> Is it possible that those intents are getting alternately written and deleted?

I don't think so because the timestamp on the intents aren't changing. I added more debug logging around these intents, so we'll see what pops out when the cluster gets in this state again.

**Comment by bdarnell** [comment]
> I don't think so because the timestamp on the intents aren't changing.

The timestamp in the TxnMeta is not changing. The timestamp on the MVCCKey could still be (and in fact this is what happens when there is a WriteTooOldError: The MVCC intent is written at the new timestamp but the original one is still written in the TxnMeta. Subsequent epochs of the transaction should be getting their timestamps bumped anyway, but I think it's at least theoretically possible that the bumped transaction could be writing at intents at higher timestamps while the TxnMeta stays in the past.

**Comment by bdarnell** [comment]
To clarify a little more about timestamps: the timestamp is actually stored *three* times: MVCCMetadata.Timestamp, MVCCMetadata.Txn.Timestamp, and in the versioned key. It's normal for MVCCMetadata.Txn.Timestamp to differ from the other two, but if MVCCMetadata.Timestamp ever differed from the timestamp in the versioned key for any intent, chaos would result (the intent would not be correctly removed on abort. I don't think we have any evidence of this happening, unless it's behind #30792).

**Comment by bdarnell** [comment]
DIscussion of this issue has migrated for the moment to https://github.com/cockroachdb/cockroach/issues/28358#issuecomment-435967814

**Comment by bdarnell** [comment]
OK, we have another repro of this on adriatic. All processes are currently stopped to allow investigation. This is a checkpoint of what I've found so far. 

I haven't found an exact occurrence of the bug yet (I'm not sure where to look for the keys that are triggering errors while the cluster is down), but I've found the key elements of it in 326447.sst on adriatic-0004. I arrived at this file because I noticed that our local-key encoding leads to pathological compaction behavior since approximately every L0 sstable includes both range-local and regular keys, which spans a large number of sstables in lower levels when it gets compacted. Any mid-level sst containing range-local keys gets compacted all the time, so the most recent compactions on several nodes contained repeated compactions following a pattern like this (from the tail of node 4's manifest):

```
VersionEdit {
  LogNumber: 326439
  PrevLogNumber: 0
  NextFileNumber: 326441
  LastSeq: 16092582865
  AddFile: 0 326440 13935919 '^Ai<89>rrask
VersionEdit {
  LogNumber: 326441
  PrevLogNumber: 0
  NextFileNumber: 326443
  LastSeq: 16093235319
  AddFile: 0 326442 7372183 '^Ai<89>rrask
VersionEdit {
  PrevLogNumber: 0
  NextFileNumber: 326447
  LastSeq: 16093242823
  DeleteFile: 0 326440
  DeleteFile: 0 326442
  DeleteFile: 3 326422
  DeleteFile: 3 326424
  DeleteFile: 3 326425
  AddFile: 3 326443 17025581 '^Ai<89>rrask
VersionEdit {
  PrevLogNumber: 0
  NextFileNumber: 326451
  LastSeq: 16093256508
  DeleteFile: 3 326443
  DeleteFile: 4 326366
  DeleteFile: 4 326368
  DeleteFile: 4 326369
  AddFile: 4 326447 15090662 '^Ai<89>rrask
```

326447.sst is an L4 sstable, created by merging an L3 sstable and 3 L4's. The L3 sstable was just written prior to this compaction as a combination of two just-written L0's and 3 more L3's. This sst is about 80% range-local keys (and contains nearly all the range-local keys for this node). The relatively low content of "normal" keys may contribute to the sst having a narrow timestamp span (a normal-only sst would have a higher chance of extending its timestamp span to cover its lowest-sequence-number entries)

```
root@cockroach-adriatic-0004:~# ~/cockroach debug sst_dump --file=/mnt/data1/326447.sst --output_hex --input_key_hex --show_properties
from [] to []
Process /mnt/data1/326447.sst
Sst file format: block-based
Table Properties:
------------------------------
  # data blocks: 666
  # entries: 175709
  # range deletions: 0
  raw key size: 4654773
  raw average key size: 26.491375
  raw value size: 19255116
  raw average value size: 109.585257
  data block size: 14882170
  index block size: 25085
  filter block size: 189509
  (estimated) table size: 15096764
  filter policy name: rocksdb.BuiltinBloomFilter
  column family ID: 0
  column family name: default
  comparator name: cockroach_comparator
  merge operator name: cockroach_merge_operator
  property collectors names: [TimeBoundTblPropCollectorFactory]
  SST file compression algo: Snappy
  creation time: 1541905947
  time stamp of earliest key: 0
  # deleted keys: 130606
  # merge operands: 30
Raw user collected properties
------------------------------
  # crdb.ts.max: 0x1565F349B4146C54
  # crdb.ts.min: 0x1565F26ED305C648
  # rocksdb.block.based.table.index.type: 0x00000000
  # rocksdb.block.based.table.prefix.filtering: 0x31
  # rocksdb.block.based.table.whole.key.filtering: 0x30
  # rocksdb.deleted.keys: 0xAEFC07
  # rocksdb.merge.operands: 0x1E
```

This sst spans from rocksdb sequence numbers 16088995015 to 16093230788. Key BF898C898F8800 is an intent on /Table/55/1/4/1/7/0. It was deleted in this sst at sequence number 16088995015 (i.e. the oldest sequence number in the file), but no other keys with the same prefix exist, so the timestamped value for this intent must exist in some older sst, which is the hypothesized cause of the bug. 

However, when we look for the previous write to this key, it looks like it has a timestamp that falls within the range given in the table properties. 

```
for i in /mnt/data1/*.sst; do ~/cockroach debug sst_dump --file=$i --output_hex --input_key_hex --command=scan --from=0xBF898C898F8800 --to=0xBF898C898F8801; done
...
Process /mnt/data1/326411.sst
Sst file format: block-based
'BF898C898F8800' seq:16088994721, type:1 => 0A2A0A1082A7BF2BAA5641D9A4FE5A81520672FA1A04BD898C882A0A08C4C4C08CF2CDFCB21530B1BC193808120C08C4C4C08CF2CDFCB21510001800200C28D203
'BF898C898F88001565F26F2190224409' seq:16088994722, type:1 => ACC83A720A460977677635676A6E526816024F45160A424152424152414E5449160D3837534D666856434B527850321612646C65417943676837375775335945354435160A76327572325A4564483616024543160935323839313131313116103538393537353936323933363834393818CA83CABB0800160247431505348D4C4B4015032904321505348D5783B41506348E034E7F0C13B80313A003168233356170704F633557464D51667053465431415139726C535944563844684D4F4D526F53786A31466B50336F6F74733154466966677135325A784C6F5864356F4B49796F4238735571483043796F6354704A46554E63377871594F38776B6E434A44696663317579656254466E615064636D53437535574F645864774B4476667142366B5A6D7049555546647371654A376161347244527A4B346B734D65593141636F5261556364736E6B716A526C47465779345551664D4732646D455465584A63456F42414F4770326131444B6D5369454F4D35683676657136536578553758746951337A5171534E69395035536E347075466D4C454C42765A737064786C6A5777755533564A376C6573704349673257536B725830766964356D63774B4771754D545976766279576A72366D794675744B44
...
```

That value's timestamp is 1565F26F21902244, which falls in between 0x1565F26ED305C648 and 0x1565F349B4146C54. So this isn't our problematic key. However, there are other intent deletions in this sst that don't have corresponding timestamped values (and relatively low sequence numbers), so we've proven that the major elements of the bug exist. I'll test some of those other keys tomorrow.

**Comment by bdarnell** [comment]
> I'll test some of those other keys tomorrow.

No luck. This sst contains 10 tombstones for metadata keys without corresponding values, but in each case it appears that the value happens to fall within the bounds discovered by the property collector. 

In any case, this confirms that metadata tombstones which convey no information to the time bound property collector are pretty common, and it's just a matter of luck that there are usually other KV entries that allow us to set bounds that are wide enough (and that sstables in higher levels tend to be short-lived and compact away before the bug is triggered to larger lower levels with wider bounds, or all the way to L6 with no tombstones).

Are there any more questions that would be answered by digging deeper on this, or are we ready to move on to figuring out a solution?

**Comment by tbg** [comment]
@nvanbenschoten should we reopen this or is there a new issue tracking the current QPS stalls?

**Comment by nvanbenschoten** [comment]
I don't think we should re-open this because the symptoms were completely different. I opened #32582 to track the new issue.
